File: internal/store/observation.go
package store

// observation.go defines the ObservationKind taxonomy — a closed, semantic set
// of categories derived from EventKind. It is the single source of truth for
// "what kind of thing is this event" and is consumed by the runstatus SPA and
// any other display layer.
//
// The mapping is a pure total function: same input, same output, no I/O. It
// adds no field to the on-disk event format (the category is computed at read
// time). The companion test in observation_test.go guards against drift when
// new EventKind constants are added.

// Kind is the observation-kind taxonomy — the closed set of semantic categories
// that every EventKind maps to.
type Kind string

const (
	// KindDecision covers interpretive choices with available/chosen/confidence.
	// These are the primary "moat" events: gate resolutions and off-path Q&A.
	KindDecision Kind = "decision"

	// KindRouting covers events that advance the turn or record how it was routed.
	KindRouting Kind = "routing"

	// KindAgentCall covers LLM/operator calls with prompt, response, cost and latency.
	KindAgentCall Kind = "agent-call"

	// KindHostCall covers deterministic side-effecting host execution.
	KindHostCall Kind = "host-call"

	// KindNarration covers operator-facing text: room narration and rendered views.
	KindNarration Kind = "narration"

	// KindWorldMutation covers set: world writes.
	KindWorldMutation Kind = "world-mutation"

	// KindLifecycle covers structural and bookkeeping events that do not fit
	// the above categories.
	KindLifecycle Kind = "lifecycle"
)

// ObservationKind returns the semantic ObservationKind for a given EventKind.
// The function is a pure total function: every declared EventKind maps to
// exactly one non-empty Kind, and any unrecognised string falls back to
// KindLifecycle. It never returns an empty string.
func ObservationKind(kind EventKind) Kind {
	switch kind {
	// decision — interpretive choices with available/chosen/confidence
	case GateDecided, WriteModeGranted, OffPathQuestion, OffPathAnswer:
		return KindDecision

	// routing — what advanced the turn and how it was routed
	case TurnStarted, IntentAccepted:
		return KindRouting

	// agent-call — LLM/operator calls with prompt/response/cost/latency
	case AgentCalled, AgentReturned, AgentError, LLMToolCall:
		return KindAgentCall

	// host-call — deterministic side-effecting execution
	case HostInvoked, HostDispatched, HostReturned, HarnessError:
		return KindHostCall

	// narration — operator-facing text
	case MachineSay, TurnEnded:
		return KindNarration

	// world-mutation — set: world writes
	case EffectApplied:
		return KindWorldMutation

	// lifecycle — structural/bookkeeping
	case ValidationFailed, TransitionApplied, StateExited, StateEntered,
		GuardRejected, JobSubmitted, JobCompleted, TimeoutFired, MachineError,
		OffPathEntered, OffPathExited, IDEContextCaptured,
		StorySnapshot, StoryChanged, UserInputReceived:
		return KindLifecycle

	default:
		// Fallback: unknown kinds are lifecycle. The test in observation_test.go
		// asserts every declared constant is explicitly handled, so this branch
		// fires only for EventKind values not yet declared.
		return KindLifecycle
	}
}

File: tools/runstatus/src/lib/observation.ts
/**
 * observation.ts — semantic taxonomy over EventKind strings.
 *
 * Mirrors internal/store/observation.go exactly. The mapping is a pure total
 * function: same input, same output. The Go test (observation_test.go) is the
 * canonical drift guard; the Vitest counterpart asserts equality against the
 * Go golden dump.
 *
 * No new event field is added. The category is derived at read time.
 */

export type ObservationKind =
  | "decision"
  | "agent-call"
  | "host-call"
  | "narration"
  | "world-mutation"
  | "routing"
  | "lifecycle";

/**
 * observationKind returns the semantic category for a given event msg string.
 * Falls back to "lifecycle" for any unrecognised string — never returns empty.
 */
export function observationKind(msg: string): ObservationKind {
  switch (msg) {
    // decision — interpretive choices with available/chosen/confidence
    case "machine.gate_decided":
    case "machine.write_mode_granted":
    case "agent.off_path.question":
    case "agent.off_path.answer":
      return "decision";

    // routing — what advanced the turn and how it was routed
    case "turn.start":
    case "machine.intent_accepted":
      return "routing";

    // agent-call — LLM/operator calls with prompt/response/cost/latency
    case "agent.call.start":
    case "agent.call.complete":
    case "agent.call.error":
    case "agent.tool_call":
      return "agent-call";

    // host-call — deterministic side-effecting execution
    case "harness.called":
    case "harness.dispatched":
    case "harness.returned":
    case "harness.error":
      return "host-call";

    // narration — operator-facing text
    case "machine.say":
    case "turn.end":
      return "narration";

    // world-mutation — set: world writes
    case "world.update":
      return "world-mutation";

    // lifecycle — structural/bookkeeping (and all unknown strings)
    default:
      return "lifecycle";
  }
}

/**
 * COLOR_MAP maps each ObservationKind to a canonical hex color.
 * Used by TraceTimeline category chips and per-row observation dots.
 */
export const COLOR_MAP: Record<ObservationKind, string> = {
  decision:       "#7c3aed", // purple
  "agent-call":  "#0ea5e9", // blue
  "host-call":    "#f59e0b", // amber
  narration:      "#10b981", // emerald
  "world-mutation": "#ec4899", // pink
  routing:        "#6366f1", // indigo
  lifecycle:      "#475569", // slate
};
